File: VideoMergeIOS/VideoMerger/UI/MergeViewModel.swift
//
//  MergeViewModel.swift
//  VideoMerger
//
//  iOS 移植自 Android VideoMerger 的 MergeViewModel.kt
//

import Foundation
import SwiftUI
import Combine
import Photos

/// UI 状态
struct MergeUiState: Equatable {
    var selectedVideoURLs: [URL] = []
    var mergeType: MergeType = .grid
    var options: MergeOptions = MergeOptions()
    var isProcessing: Bool = false
    var progress: Float = 0
    var statusMessage: String = ""
    var logLines: String = ""
    var mergeResult: MergeResult? = nil
    var isSaved: Bool = false
    var isSaving: Bool = false
    var isFullscreen: Bool = false
    var fullscreenVideoURL: URL? = nil
    var errorMessage: String? = nil
    var history: [HistoryEntry] = []
    var showHistory: Bool = false
}

@MainActor
final class MergeViewModel: ObservableObject {

    @Published var uiState = MergeUiState()

    private let engine = MergeEngine()

    init() {
        uiState.history = engine.loadHistory()
    }

    // MARK: - 选择视频

    func onVideosSelected(_ urls: [URL]) {
        uiState.selectedVideoURLs = urls
        uiState.errorMessage = nil
        uiState.mergeResult = nil
        uiState.isSaved = false
        uiState.logLines = ""
    }

    /// 追加视频（预览卡缩略图条末尾的「+」按钮）。
    /// 按标准化路径去重，避免重复选择同一视频；输入变化后旧结果作废。
    func addVideos(_ urls: [URL]) {
        guard !urls.isEmpty else { return }
        var existing = Set(uiState.selectedVideoURLs.map { $0.standardizedFileURL.absoluteString })
        let fresh = urls.filter { !existing.contains($0.standardizedFileURL.absoluteString) }
        guard !fresh.isEmpty else { return }
        existing.formUnion(fresh.map { $0.standardizedFileURL.absoluteString })
        uiState.selectedVideoURLs += fresh
        uiState.errorMessage = nil
        uiState.mergeResult = nil
        uiState.isSaved = false
        uiState.logLines = ""
    }

    /// 移除单个已选视频（缩略图条右上角「−」按钮）。
    /// 输入变化后旧结果作废；列表清空时顺带清掉错误提示。
    func removeVideo(at index: Int) {
        guard uiState.selectedVideoURLs.indices.contains(index) else { return }
        uiState.selectedVideoURLs.remove(at: index)
        if uiState.selectedVideoURLs.isEmpty {
            uiState.errorMessage = nil
        }
        uiState.mergeResult = nil
        uiState.isSaved = false
        uiState.logLines = ""
    }

    // MARK: - 合并类型 & 选项

    func onMergeTypeSelected(_ type: MergeType) {
        uiState.mergeType = type
        // 照片墙是随机布局：首次进入时固定一个种子，预览与导出才会是同一版式
        if type == .photoWall && uiState.options.photoWallSeed == nil {
            uiState.options.photoWallSeed = newPhotoWallSeed()
        }
    }

    /// 照片墙「换一批」：换种子即换版式，预览与导出同步。
    func shufflePhotoWall() {
        uiState.options.photoWallSeed = newPhotoWallSeed()
    }

    private func newPhotoWallSeed() -> Int {
        Int.random(in: 1..<Int.max)
    }

    func updateOptions(_ options: MergeOptions) {
        uiState.options = options
    }

    // MARK: - 启动合并

    func startMerge() {
        if uiState.selectedVideoURLs.isEmpty {
            uiState.errorMessage = "请先选择视频"
            return
        }
        if uiState.selectedVideoURLs.count == 1 {
            uiState.errorMessage = "至少需要选择 2 个视频"
            return
        }

        uiState.isProcessing = true
        uiState.progress = 0
        uiState.statusMessage = "准备中..."
        uiState.errorMessage = nil
        uiState.mergeResult = nil
        uiState.isSaved = false
        uiState.logLines = ""

        let urls = uiState.selectedVideoURLs
        let mergeType = uiState.mergeType
        let options = uiState.options

        engine.merge(
            videoURLs: urls,
            mergeType: mergeType,
            options: options,
            onProgress: { [weak self] progress in
                let msg: String
                if progress < 0.1 { msg = "复制视频文件中..." }
                else if progress < 0.98 { msg = "合并中... \(Int(progress * 100))%" }
                else { msg = "正在完成..." }
                self?.uiState.progress = progress
                self?.uiState.statusMessage = msg
            },
            onLog: { [weak self] line in
                guard let self = self else { return }
                // 限制日志总长度（保留末尾 8KB），避免 Text 渲染过多内容卡 UI
                let combined = self.uiState.logLines + line
                let maxLen = 8 * 1024
                if combined.count > maxLen {
                    let start = combined.index(combined.endIndex, offsetBy: -maxLen)
                    self.uiState.logLines = "…(已截断)\n" + combined[start...]
                } else {
                    self.uiState.logLines = combined
                }
            },
            completion: { [weak self] result in
                switch result {
                case .success(let mergeResult):
                    self?.uiState.isProcessing = false
                    self?.uiState.progress = 1
                    self?.uiState.statusMessage = "合并完成！预览中"
                    self?.uiState.mergeResult = mergeResult
                    AppAnalytics.mergeExportSuccess(mergeResult.mergeType)
                case .failure(let err):
                    self?.uiState.isProcessing = false
                    self?.uiState.progress = 0
                    self?.uiState.statusMessage = ""
                    self?.uiState.errorMessage = err.localizedDescription
                }
            }
        )
    }

    // MARK: - 保存

    func saveResult() {
        guard let result = uiState.mergeResult else { return }
        uiState.isSaving = true
        engine.saveResult(result) { [weak self] res in
            switch res {
            case .success:
                self?.uiState.isSaving = false
                self?.uiState.isSaved = true
                self?.uiState.history = self?.engine.loadHistory() ?? []
            case .failure(let err):
                self?.uiState.isSaving = false
                self?.uiState.errorMessage = "保存失败: \(err.localizedDescription)"
            }
        }
    }

    /// 把任意视频（含历史记录里的成片）保存到相册。
    /// ★ 不能用 saveResult()：那条路径 guard mergeResult != nil，
    ///   从历史页点开视频时 mergeResult 为 nil，会静默无反应。
    func saveVideoToGallery(url: URL) {
        uiState.isSaving = true
        engine.saveExistingVideoToGallery(url: url) { [weak self] res in
            switch res {
            case .success:
                self?.uiState.isSaving = false
                self?.uiState.isSaved = true
                self?.uiState.statusMessage = "已保存到相册"
            case .failure(let err):
                self?.uiState.isSaving = false
                self?.uiState.errorMessage = "保存失败: \(err.localizedDescription)"
            }
        }
    }

    // MARK: - 全屏播放

    func openFullscreen() {
        guard let url = uiState.mergeResult?.outputFileURL else { return }
        uiState.isFullscreen = true
        uiState.fullscreenVideoURL = url
    }

    func openFullscreen(url: URL) {
        uiState.isFullscreen = true
        uiState.fullscreenVideoURL = url
    }

    func closeFullscreen() {
        uiState.isFullscreen = false
        uiState.fullscreenVideoURL = nil
    }

    // MARK: - 清除

    func clearResult() {
        MediaUtils.cleanupOutputFiles()
        uiState.mergeResult = nil
        uiState.isSaved = false
        uiState.logLines = ""
    }

    func clearAll() {
        MediaUtils.cleanupOutputFiles()
        uiState.selectedVideoURLs = []
        uiState.isProcessing = false
        uiState.progress = 0
        uiState.statusMessage = ""
        uiState.logLines = ""
        uiState.mergeResult = nil
        uiState.isSaved = false
        uiState.isSaving = false
        uiState.errorMessage = nil
    }

    // MARK: - 历史

    func showHistoryPage() {
        uiState.history = engine.loadHistory()
        uiState.showHistory = true
    }

    func hideHistoryPage() {
        uiState.showHistory = false
    }

    func dismissError() {
        uiState.errorMessage = nil
    }

    func deleteHistoryEntry(_ entry: HistoryEntry) {
        engine.deleteHistoryEntry(entry: entry)
        uiState.history = engine.loadHistory()
    }
}

File: VideoMergeAndroid/app/src/main/java/com/moon/videomerger/ui/MergeViewModel.kt
package com.moon.videomerger.ui

import android.app.Application
import android.net.Uri
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.moon.videomerger.merger.MergeEngine
import com.moon.videomerger.merger.MergeOptions
import com.moon.videomerger.merger.MergeResult
import com.moon.videomerger.merger.MergeType
import com.moon.videomerger.util.MediaUtils
import com.moon.videomerger.util.VideoHistoryStore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlin.random.Random

/**
 * UI 状态
 */
data class MergeUiState(
    val selectedVideos: List<Uri> = emptyList(),
    val mergeType: MergeType = MergeType.GRID,
    val options: MergeOptions = MergeOptions(),
    val isProcessing: Boolean = false,
    val progress: Float = 0f,
    val statusMessage: String = "",
    val logLines: String = "",
    val mergeResult: MergeResult? = null,
    val isSaved: Boolean = false,
    val isSaving: Boolean = false,
    val isFullscreen: Boolean = false,
    val fullscreenVideoPath: String? = null,
    val errorMessage: String? = null,
    val history: List<VideoHistoryStore.HistoryEntry> = emptyList(),
    val showHistoryDialog: Boolean = false
)

class MergeViewModel(app: Application) : AndroidViewModel(app) {

    private val engine = MergeEngine(app.applicationContext)

    private val _uiState = MutableStateFlow(MergeUiState())
    val uiState: StateFlow<MergeUiState> = _uiState.asStateFlow()

    init {
        // 加载本地历史
        val history = engine.loadHistory()
        _uiState.value = _uiState.value.copy(history = history)
    }

    fun onVideosSelected(uris: List<Uri>) {
        _uiState.value = _uiState.value.copy(
            selectedVideos = uris,
            errorMessage = null,
            mergeResult = null,
            isSaved = false,
            logLines = ""
        )
    }

    /**
     * 追加视频（预览卡缩略图条末尾的「+」按钮）。
     * 按 Uri 字符串去重，避免重复选择同一视频；输入变化后旧结果作废。
     */
    fun addVideos(uris: List<Uri>) {
        if (uris.isEmpty()) return
        val existing = _uiState.value.selectedVideos.map { it.toString() }.toSet()
        val fresh = uris.filter { it.toString() !in existing }
        if (fresh.isEmpty()) return
        _uiState.value = _uiState.value.copy(
            selectedVideos = _uiState.value.selectedVideos + fresh,
            errorMessage = null,
            mergeResult = null,
            isSaved = false,
            logLines = ""
        )
    }

    /**
     * 移除单个已选视频（缩略图条右上角「−」按钮）。
     * 输入变化后旧结果作废；列表清空时顺带清掉错误提示。
     */
    fun removeVideo(index: Int) {
        val list = _uiState.value.selectedVideos.toMutableList()
        if (index !in list.indices) return
        list.removeAt(index)
        _uiState.value = _uiState.value.copy(
            selectedVideos = list,
            errorMessage = if (list.isEmpty()) null else _uiState.value.errorMessage,
            mergeResult = null,
            isSaved = false,
            logLines = ""
        )
    }

    fun onMergeTypeSelected(type: MergeType) {
        val state = _uiState.value
        // 照片墙是随机布局：首次进入时固定一个种子，预览与导出才会是同一版式
        val options = if (type == MergeType.PHOTO_WALL && state.options.photoWallSeed == null) {
            state.options.copy(photoWallSeed = newPhotoWallSeed())
        } else {
            state.options
        }
        _uiState.value = state.copy(mergeType = type, options = options)
    }

    /** 照片墙「换一批」：换种子即换版式，预览与导出同步。 */
    fun shufflePhotoWall() {
        _uiState.value = _uiState.value.copy(
            options = _uiState.value.options.copy(photoWallSeed = newPhotoWallSeed())
        )
    }

    private fun newPhotoWallSeed(): Int = Random.nextInt(1, Int.MAX_VALUE)

    fun updateOptions(options: MergeOptions) {
        _uiState.value = _uiState.value.copy(options = options)
    }

    fun startMerge() {
        val state = _uiState.value
        if (state.selectedVideos.isEmpty()) {
            _uiState.value = state.copy(errorMessage = "请先选择视频")
            return
        }
        if (state.selectedVideos.size == 1) {
            _uiState.value = state.copy(errorMessage = "至少需要选择 2 个视频")
            return
        }

        viewModelScope.launch {
            _uiState.value = state.copy(
                isProcessing = true,
                progress = 0f,
                statusMessage = "准备中...",
                errorMessage = null,
                mergeResult = null,
                isSaved = false,
                logLines = ""
            )

            val result = engine.merge(
                videoUris = state.selectedVideos,
                mergeType = state.mergeType,
                options = state.options,
                onProgress = { progress ->
                    val msg = when {
                        progress < 0.1f -> "复制视频文件中..."
                        progress < 0.98f -> "合并中... ${(progress * 100).toInt()}%"
                        else -> "生成缩略图..."
                    }
                    _uiState.value = _uiState.value.copy(
                        progress = progress,
                        statusMessage = msg
                    )
                },
                onLog = { line ->
                    _uiState.value = _uiState.value.copy(
                        logLines = _uiState.value.logLines + line
                    )
                }
            )

            result.fold(
                onSuccess = { mergeResult ->
                    _uiState.value = _uiState.value.copy(
                        isProcessing = false,
                        progress = 1f,
                        statusMessage = "合并完成！预览中",
                        mergeResult = mergeResult
                    )
                },
                onFailure = { e ->
                    _uiState.value = _uiState.value.copy(
                        isProcessing = false,
                        progress = 0f,
                        statusMessage = "",
                        errorMessage = e.message ?: "未知错误"
                    )
                }
            )
        }
    }

    /**
     * 保存到相册 + 本地历史。
     */
    fun saveResult() {
        val result = _uiState.value.mergeResult ?: return
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isSaving = true)
            val saveResult = engine.saveResult(result)
            saveResult.fold(
                onSuccess = {
                    // 刷新历史
                    val history = engine.loadHistory()
                    _uiState.value = _uiState.value.copy(
                        isSaving = false,
                        isSaved = true,
                        history = history
                    )
                },
                onFailure = { e ->
                    _uiState.value = _uiState.value.copy(
                        isSaving = false,
                        errorMessage = "保存失败: ${e.message}"
                    )
                }
            )
        }
    }

    /**
     * 全屏播放合并结果。
     */
    fun openFullscreen() {
        val path = _uiState.value.mergeResult?.outputFile?.absolutePath ?: return
        _uiState.value = _uiState.value.copy(
            isFullscreen = true,
            fullscreenVideoPath = path
        )
    }

    /**
     * 全屏播放历史视频。
     */
    fun openFullscreen(path: String) {
        _uiState.value = _uiState.value.copy(
            isFullscreen = true,
            fullscreenVideoPath = path
        )
    }

    fun closeFullscreen() {
        _uiState.value = _uiState.value.copy(
            isFullscreen = false,
            fullscreenVideoPath = null
        )
    }

    /**
     * 清除合并结果，回到初始状态，并清理输出文件。
     */
    fun clearResult() {
        MediaUtils.cleanupOutputFiles(getApplication())
        _uiState.value = _uiState.value.copy(
            mergeResult = null,
            isSaved = false,
            logLines = ""
        )
    }

    /**
     * 清除全部操作，恢复初始状态。
     */
    fun clearAll() {
        MediaUtils.cleanupOutputFiles(getApplication())
        _uiState.value = _uiState.value.copy(
            selectedVideos = emptyList(),
            isProcessing = false,
            progress = 0f,
            statusMessage = "",
            logLines = "",
            mergeResult = null,
            isSaved = false,
            isSaving = false,
            errorMessage = null
        )
    }

    fun showHistory() {
        val history = engine.loadHistory()
        _uiState.value = _uiState.value.copy(history = history, showHistoryDialog = true)
    }

    fun hideHistory() {
        _uiState.value = _uiState.value.copy(showHistoryDialog = false)
    }

    fun dismissError() {
        _uiState.value = _uiState.value.copy(errorMessage = null)
    }

    fun deleteHistoryEntry(entry: VideoHistoryStore.HistoryEntry) {
        engine.deleteHistoryEntry(entry)
        val history = engine.loadHistory()
        _uiState.value = _uiState.value.copy(history = history)
    }

    override fun onCleared() {
        super.onCleared()
        MediaUtils.cleanupOutputFiles(getApplication())
    }
}
